cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java

```diff
@@ -0,0 +1,177 @@
+        try (MetricsTable v3MetricsTable = migrationTableUtility.getV3MetricsTable(resolution);
+             MetricsTable v2MetricsTable = migrationTableUtility.getV2MetricsTable(resolution)) {
+          if (v2MetricsTable == null || v3MetricsTable == null) {
+            // dataset/hbase service not available, skipping check this time, will try in next schedule.
+            // return false to try again in next attempt,
+            // by false we will proceed to run the migration, however if all tables have been migrated
+            // migrationTask can handle that scenario and shutdown the executor service.
+            return false;
+          }
+          MetricsTableMigration metricsTableMigration =
+            new MetricsTableMigration(v2MetricsTable, v3MetricsTable, cConf);
+          // if metrics data is available in old table, return false
+          if (metricsTableMigration.isOldMetricsDataAvailable()) {
+            return false;
+          }
+        } catch (IOException e) {
+          // exception while closing metrics tables - ignoring
+        }
+      }
+    }
+    // all tables data have been migrated, so migration is complete
+    return true;
+  }
+
+  private final class MigrationTask implements Runnable {
+    public void run() {
+
+      Iterator<Integer> resolutions = maxRecordsToScanResolutionMap.keySet().iterator();
+      // iterate through resolutions, remove entry from iterator
+      // if the table has been deleted for that resolution or if there are no more data available in the table
+      while (resolutions.hasNext()) {
+        int resolution = resolutions.next();
+        if (migrationTableUtility.v2MetricsTableExists(resolution)) {
+          try (MetricsTable v3MetricsTable = migrationTableUtility.getV3MetricsTable(resolution);
+               MetricsTable v2MetricsTable = migrationTableUtility.getV2MetricsTable(resolution)) {
+            if (v2MetricsTable == null || v3MetricsTable == null) {
+              // dataset/hbase service not available, return,
+              // will try to get table and perform migration in next schedule.
+              return;
+            }
+
+            MetricsTableMigration metricsTableMigration =
+              new MetricsTableMigration(v2MetricsTable, v3MetricsTable, cConf);
+
+            try {
+              if (metricsTableMigration.isOldMetricsDataAvailable()) {
+                LOG.info("Metrics data is available in v2 metrics - {} resolution table.. " +
+                           "Running Migration", resolution);
+                metricsTableMigration.transferData(maxRecordsToScanResolutionMap.get(resolution));
+                break;
+              } else {
+                // no more metrics data is available for transfer, remove from iterator
+                resolutions.remove();
+              }
+            } catch (Exception e) {
+              LOG.warn("Exception while performing metrics data transfer", e);
```

Comment:
what if it is an interruptedException? That would happen when metrics processor shuts down and calls shutdownNow() for the executor. (does it? - it should)